crud/common/utils.lua

```diff
@@ -210,6 +210,35 @@ local function determine_enabled_features()
     -- since Tarantool 2.6.3 / 2.7.2 / 2.8.1
     enabled_tarantool_features.jsonpath_indexes = major >= 3 or (major >= 2 and ((minor >= 6 and patch >= 3)
         or (minor >= 7 and patch >= 2) or (minor >= 8 and patch >= 1) or minor >= 9))
+
+    -- The merger module was implemented in 2.2.1, see [1].
+    -- However it had the critical problem [2], which leads to
+    -- segfault at attempt to use the module from a fiber serving
+    -- iproto request. So we don't use it in versions before the
+    -- fix.
+    --
+    -- [1]: https://github.com/tarantool/tarantool/issues/3276
+    -- [2]: https://github.com/tarantool/tarantool/issues/4954
+    enabled_tarantool_features.builtin_merger =
+        (major == 2 and minor == 3 and patch >= 3) or
+        (major == 2 and minor == 4 and patch >= 2) or
+        (major == 2 and minor == 5 and patch >= 1) or
+        (major >= 2 and minor >= 6) or
+        (major >= 3)
```

Comment:
Some conditions above won't work for Tarantool 3.0. E.g. "enabled_tarantool_features.uuids" and "enabled_tarantool_features.fieldpaths".

Reply:
Do you want to reformat other conditions in the way I use here? I can do in a separate commit.

Reply:
Yes. I think other conditions should support next major Tarantool versions.
I suggest to do it in this patch in separate commit or even a bit later in separate PR.

Reply:
Aside of obvious problems:

* `utils.tarantool_supports_fieldpaths()` and `utils.tarantool_supports_uuids()` are incorrect for tarantool-3* (tarantool-4* and so on).
* `utils.tarantool_supports_uuids()` and `utils.tarantool_supports_jsonpath_indexes()` are used only in tests.

There is one interesting:

```lua
-- since Tarantool 2.6.3 / 2.7.2 / 2.8.1
enabled_tarantool_features.jsonpath_indexes = <...>
```

However JSON path indexes were introduced in 2.1.2, see the following links:

https://github.com/tarantool/tarantool/issues/1012
https://github.com/tarantool/tarantool/commit/4273ec52e122d6d37c8deedf1bc10732a7e40c0e
https://github.com/tarantool/tarantool/commit/a754980d7feab110b8c82ee15ef13e080afa2882

I re-checked 2.6.3, 2.7.2 and 2.8.1 release notes, there is no 'json' word in them:

https://github.com/tarantool/tarantool/releases/2.6.3
https://github.com/tarantool/tarantool/releases/2.7.2
https://github.com/tarantool/tarantool/releases/2.8.1

This code was added in PR #158 and I don't see anything that would clarify why those versions are choosen. I looked at the issue as well: #12 and...

> Support JSON path
>
> Even Tarantool 1.10 supports value extraction via jsonpath. So we could patch our filter generator and add jsonpath's support of jsonpath conditions.

It is not about JSON path indexes? It is about filtering values using filters that use JSON path notation? How '1.10 supports value extraction via jsonpath'? Or it means that we can parse such notation using Lua without tarantool side support?

(What a hell...)

AFAIU from code and tests in PR #158 there was a problem with `crud.select()` with index name in a condition, when this index is constructed by a JSON path. The problem was resolved by using the `key_def`/`tuple.keydef` module to correctly extract the key and it will not work if the module is not present.

(Ugh... Why not describe things for humans in a changelog? That's my regular toothache: convince others that development is about skill to share knowledge rather than about coding.)

If my statements are correct, then we need the following property in *tests*: whether `<keydef object>:extract_key()` supports JSON paths. If so, run the tests. Otherwise a user will receive an error (I hope, but it is not tested). Ideally we should test all possible variants of 'keydef supports json path' and 'json path index is supported' properties.

It still doesn't correspond to 2.6.3/2.7.2/2.8.1 anyhow.

I don't know what to do here, got tired, will think later.

Reply:
> It is not about JSON path indexes? It is about filtering values using filters that use JSON path notation? How '1.10 supports value extraction via jsonpath'? Or it means that we can parse such notation using Lua without tarantool side support?

There were two parts of this task.
The first is json path in tuple. This feature was introduced in 1.10 (https://github.com/tarantool/tarantool/issues/1285).
The second is json path indexes (https://github.com/tarantool/tarantool/issues/1012) - 2.1.

> It still doesn't correspond to 2.6.3/2.7.2/2.8.1 anyhow.

It looks like a typo. Of course JSONPath indexes were introduced in 2.1.

Reply:
I see now, there was PR #150. Thanks!

Reply:
I'll prepare a patchset of two commits: fix of the checks itself and move two of them to test helpers. I would detach it from this change and open as a separate PR: it has more changes than I would comfortable include here, where I want to fix the problem with merger.

Reply:
No problem. This PR LGTM. Let's push it.

Reply:
Nice, thanks!

Reply:
Just changing JSON index support condition to 2.1.2 (and so enabling relevant tests) leads to fail in CI on tarantool 2.3-2.5 (maybe 2.2 as well). [Results](https://github.com/tarantool/crud/actions/runs/1466557064). I'll return here back later.